### PR TITLE
✨Add ClusterTopologyController scaffolding

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This is part of the activities for the implementation of the [ClusterClass proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/202105256-cluster-class-and-managed-topologies.md). 

This PR creates the scaffolding for the `ClusterTopologyContoller`.
As defined in the linked issue, it introduces a new UnstructuredCachingClient, set up watches, object tracker and creates the skeleton for the reconcile loop that will be completed by follow up PRs

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/4933

### CODE REVIEW

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:33]:
```suggestion
	workersMachineDeploymentTemplates                map[string]machineDeploymentTopologyClass //nolint:structcheck
```
Just a suggestion. Could be considered more consistent with the fields above (and based on the structure of ClusterClass), but I'm not even sure if I like it better.

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:31]:
```suggestion
	controlPlaneTemplate               *unstructured.Unstructured                //nolint:structcheck
```

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:30]:
```suggestion
	infrastructureClusterTemplate             *unstructured.Unstructured                //nolint:structcheck
```

Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:30]: I assume we don't need an additional prefix here and in `clusterControlPlaneTemplate` and then it would be maybe better to call the fields more or less like the resources, i.e. `infrastructureClusterTemplate` and `controlPlaneTemplate`

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:39]:
```suggestion
	bootstrapTemplate      *unstructured.Unstructured //nolint:structcheck
	infrastructureTemplate *unstructured.Unstructured //nolint:structcheck
```
Maybe drop `machine` as machine is already in the struct and field name, so we could access it via: `clusterTopologyClass.workersMachineDeploymentTemplates[i].bootstrapTemplate`

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:46]: Similar to clusterTopologyClass in case we want to change it above

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:53]:
```suggestion
// machineDeploymentTopologyState all the objects representing the state of a managed deployment.
```
nit

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:72]:
```suggestion
// Computes the desired state of the Cluster topology.
```
nit

File: controllers/clustertopology_controller.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller.go:111]: The ClusterController waits for dependant objects to be deleted; dependant object deletion is managed by other controllers (KCP, MachineDeployment, MachineSet, Machine controllers) using a combination of finalisers/explicit delete calls or cascading delete via owner dependencies to ensure everything is cleaned up in the provider infrastructure first, then in the API server. 
In case of machined deployment templates, the current implementation is using cascading delete.

File: controllers/clustertopology_controller.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller.go:155]: This what I'm advocating in https://github.com/kubernetes-sigs/cluster-api/pull/4967#issuecomment-883377554; but I'm fine as well it this will be separated PR or if it will be included in this one depending on what merges first.

File: controllers/clustertopology_controller.go
Comment (enxebre) [inline controllers/clustertopology_controller.go:52]: why is this config needed?

File: controllers/clustertopology_controller.go
Comment (enxebre) [reply controllers/clustertopology_controller.go:111]: Also in any case this check should be negative `if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {`

File: controllers/clustertopology_controller_compute.go
Comment (enxebre) [inline controllers/clustertopology_controller_compute.go:66]: typo: `Gets the Gets`

File: controllers/clustertopology_controller.go
Comment (sbueringer) [reply controllers/clustertopology_controller.go:155]: I'm also fine with both variants. Just wanted to make sure I got it right / we're all on the same page.

File: controllers/clustertopology_controller.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller.go:111]: Fixed, thanks

Comment (sbueringer) [reply controllers/clustertopology_controller.go:111]: Got it. Just wasn't sure if (e.g.) KubeadmConfigTemplate and InfrastructureMachineTemplate are considered dependent objects of a MachineDeployment and thus also deleted (I guess because they could be used in theory in multiple clusters)

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:39]: done.
I have used infrastructureMachineTemplate so also in this case we are consistent with the kind names

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:33]: I would prefer to not add workers, it is kind of redundant (we can eventually alway re iterate later when we will actually use those types in the code).

File: controllers/clustertopology_controller.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller.go:52]: This is required when calling utilconversion.ConvertReferenceAPIContract, which is something we have to do down the line

File: controllers/clustertopology_controller.go
Comment (sbueringer) [inline controllers/clustertopology_controller.go:65]:
```suggestion
```

File: controllers/clustertopology_controller.go
Comment (sbueringer) [inline controllers/clustertopology_controller.go:73]:
```suggestion
		return errors.Wrap(err, "failed setting up with a controller manager")
```

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:36]:
```suggestion
// controlPlaneTopologyClass holds the templates required for computing the desired state of a managed control plane.
type controlPlaneTopologyClass struct {
```
nit

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:31]:
```suggestion
	controlPlaneTemplate                  controlplaneTopologyClass                 //nolint:structcheck
```

Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:31]: Saw and agree with your change instead.

Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:31]: I prefer to keep only controlPlane so it will read controlPlane.template

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
@vincepri squashed commits 😅

**Comment by srm09** [comment]
/lgtm
This is turning out nicely. Great work @fabriziopandini

File: controllers/clustertopology_controller.go
Comment (ykakarap) [reply controllers/clustertopology_controller.go:155]: Yes, the ClusterClass index is being created in https://github.com/kubernetes-sigs/cluster-api/pull/4967. 
As @fabriziopandini suggested I will wait for this PR to merge before we merge the index PR. This will also give us time to have consensus on the package for indexes discussion.

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**
